### DISCUSSION TITLE
Changing sample rate based on source?

### DISCUSSION THREAD

**Discussion by theREALdotnet** [ExistentialAudio/BlackHole#635] [Q&A]
Hi all, not knowing much about Core Audio at all, I’ve been following a recipe for looping Dirac Live Processor into the audio path from various sources/players to my USB DAC. The path starts with Blackhole (0.4.0) as universal audio destination for all apps and sources, uses the Dirac plugin for room correction and ends at the USB DAC. Holding it all together is Element 0.46.6 (I’ve also tried AU Lab, same issue as below).

This works well, but the sample rate is fixed to whatever I set Blackhole to (in Audio Midi Setup). Is there a way of having the chain above change to different samples rates based on the sources, instead of converting them? Like, for example, Roon can send content at just about any sample rate directly to the USB DAC. The DAC will happily adjust. When I try the same with Blackhole as audio device in Roon, what arrives at the Dirac plugin is whatever rate Blackhole was last set to. It doesn’t seem to want to adapt to the source sample rate.

Sorry, I’m not too familiar with audio processing terminology, I hope I explained the issue well enough. Can anyone tell me whether automatically adapting to source sample rates is possible with Blackhole?

**Comment by devinroth** [comment]
BlackHole only changes sample rates when it’s told to from either users or programs. As is the case with all audio drivers, when a sample rate changes the driver informs the Hardware Abstraction Layer (HAL) which then notifies all applications connected. If an application says “no” the sample rate will not change. In order to test make sure that no other applications are using BlackHole except Roon.

Devin

**Reply by theREALdotnet** [reply]
Thanks Devin,
there is no other application running, just Roon, and the OS has been told to use the built-in speakers for system sounds.

It seems as though it’s actually Element or AU Lab that are changing the sample rate. In either, the sample rate of Blackhole needs to be selected, it can’t be left unconfigured or set to “variable”. Both Element and AU Lab will then enforce that rate, no matter what comes from Roon. Since music is playing ok, the resampling must be happening somewhere before it hits Blackhole.

When I give Roon “exclusive access” to Blackhole then Element pops up a warning saying it’s unable to change the sample rate, and refuses to play. AU Lab will simply play nothing. If I disable exclusive access both will play again.

The problem is that don’t really want to use either, but I need them to tie Dirac Live Processor to an input stream. I can set an output device in Dirac, but no input (would be nice if I could specify Blackhole there). I can also not select Dirac as audio device in Roon.

So it seems I need some kind of glue (Element or AU Lab) to put it all together. It’s all a bit confusing and frustrating at the same time.

I might check with the Element developers what they reckon.

Cheers,
Steffen.
